File: day-12/part-1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

if $PROGRAM_NAME == __FILE__
  adjacency_list = File.read(ARGV.fetch(0))
                       .lines(chomp: true)
                       .map { _1.split('-') }
                       .flat_map { [_1, _1.reverse] }
                       .group_by(&:first)
                       .transform_values { _1.map(&:last) }

  paths = []
  pending_paths = [['start']]

  while pending_paths.any?
    pending_paths = pending_paths.flat_map do |path|
      frontier = adjacency_list.fetch(path.last, [])
      frontier.filter_map do |cave|
        small_cave = cave.downcase == cave
        if cave == 'end'
          paths << path + [cave]
          nil
        elsif !small_cave || !path.include?(cave)
          path + [cave]
        end
      end
    end
  end

  puts paths.length
end

File: day-12/part-1.swift
#!/usr/bin/env swift

import Foundation

let connectedCaves = try! String(contentsOfFile: CommandLine.arguments[1])
  .split(separator: "\n")
  .map { $0.split(separator: "-").map { String($0) } }
  .flatMap { [$0, $0.reversed()] }

let adjacencyList = Dictionary(grouping: connectedCaves) { $0[0] }
  .mapValues { $0.map { $0[1] } }

var pathCount = 0
var pendingPaths = [["start"]]

while !pendingPaths.isEmpty {
  pendingPaths = pendingPaths.flatMap { path in
    adjacencyList[path.last!, default: []].compactMap { cave in
      let smallCave = cave == cave.lowercased()
      if cave == "end" {
        pathCount += 1
      } else if !smallCave || !path.contains(cave) {
        return path + [cave]
      }

      return nil
    }
  }
}

print(pathCount)
